src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -760,69 +749,54 @@ PeerImp::doAccept()
     // XXX Set timer: connection idle (idle may vary depending on connection
     // type.)
 
-    onWriteResponse(error_code(), 0);
-}
-
-http_response_type
-PeerImp::makeResponse(
-    bool crawl,
-    http_request_type const& req,
-    beast::IP::Address remote_ip,
-    uint256 const& sharedValue)
-{
-    http_response_type resp;
-    resp.result(boost::beast::http::status::switching_protocols);
-    resp.version(req.version());
-    resp.insert("Connection", "Upgrade");
-    resp.insert("Upgrade", to_string(protocol_));
-    resp.insert("Connect-As", "Peer");
-    resp.insert("Server", BuildInfo::getFullVersionString());
-    resp.insert("Crawl", crawl ? "public" : "private");
-    if (req["X-Offer-Compression"] == "lz4" && app_.config().COMPRESSION)
-        resp.insert("X-Offer-Compression", "lz4");
-
-    buildHandshake(
-        resp,
-        sharedValue,
-        overlay_.setup().networkID,
-        overlay_.setup().public_ip,
-        remote_ip,
-        app_);
-
-    return resp;
-}
-
-// Called repeatedly to send the bytes in the response
-void
-PeerImp::onWriteResponse(error_code ec, std::size_t bytes_transferred)
-{
-    if (!socket_.is_open())
-        return;
-    if (ec == boost::asio::error::operation_aborted)
-        return;
-    if (ec)
-        return fail("onWriteResponse", ec);
-    if (auto stream = journal_.trace())
-    {
-        if (bytes_transferred > 0)
-            stream << "onWriteResponse: " << bytes_transferred << " bytes";
-        else
-            stream << "onWriteResponse";
-    }
-
-    write_buffer_.consume(bytes_transferred);
-    if (write_buffer_.size() == 0)
-        return doProtocolStart();
-
-    stream_.async_write_some(
-        write_buffer_.data(),
-        bind_executor(
-            strand_,
-            std::bind(
-                &PeerImp::onWriteResponse,
-                shared_from_this(),
-                std::placeholders::_1,
-                std::placeholders::_2)));
+    auto write_buffer = [this, sharedValue]() {
+        auto buf = std::make_shared<boost::beast::multi_buffer>();
+
+        http_response_type resp;
+        resp.result(boost::beast::http::status::switching_protocols);
+        resp.version(request_.version());
+        resp.insert("Connection", "Upgrade");
+        resp.insert("Upgrade", to_string(protocol_));
+        resp.insert("Connect-As", "Peer");
+        resp.insert("Server", BuildInfo::getFullVersionString());
+        resp.insert(
+            "Crawl",
+            overlay_.peerFinder().config().peerPrivate ? "private" : "public");
+
+        if (request_["X-Offer-Compression"] == "lz4" &&
+            app_.config().COMPRESSION)
+            resp.insert("X-Offer-Compression", "lz4");
```

Comment:
It would be better to have the response set in a separate standalone function so that the handshake can be unit-tested. Though this is implemented in this PR anyways: https://github.com/ripple/rippled/pull/3644.

Reply:
I agree with you, but since you're already doing this in #3644 I think we can solve this issue when merging.